ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/translator/ASTConverter.java

```diff
@@ -160,17 +160,27 @@ private ASTNode convert() throws CalciteSemanticException {
       }
 
       HiveAggregate hiveAgg = (HiveAggregate) groupBy;
-      for (int pos : hiveAgg.getAggregateColumnsOrder()) {
-        RexInputRef iRef = new RexInputRef(groupBy.getGroupSet().nth(pos),
-            groupBy.getCluster().getTypeFactory().createSqlType(SqlTypeName.ANY));
-        b.add(iRef.accept(new RexVisitor(schema, false, root.getCluster().getRexBuilder())));
-      }
-      for (int pos = 0; pos < groupBy.getGroupCount(); pos++) {
-        if (!hiveAgg.getAggregateColumnsOrder().contains(pos)) {
+      if (hiveAgg.getAggregateColumnsOrder() != null) {
+        // Aggregation columns may have been sorted in specific order
+        for (int pos : hiveAgg.getAggregateColumnsOrder()) {
           RexInputRef iRef = new RexInputRef(groupBy.getGroupSet().nth(pos),
               groupBy.getCluster().getTypeFactory().createSqlType(SqlTypeName.ANY));
           b.add(iRef.accept(new RexVisitor(schema, false, root.getCluster().getRexBuilder())));
         }
+        for (int pos = 0; pos < groupBy.getGroupCount(); pos++) {
+          if (!hiveAgg.getAggregateColumnsOrder().contains(pos)) {
```

Comment:
Can we change this to `hiveAgg.getAggregateColumnsOrder == null || !hiveAgg.getAggregateColumnsOrder().contains(pos`) ?. Take this for loop off if condition and get rid of the repetitive else block

Reply:
That will remove the duplicate code at the cost of checking if condition within the loop. Instead, I will create an aux method with the following code:
```
          RexInputRef iRef = new RexInputRef(i,
              groupBy.getCluster().getTypeFactory().createSqlType(SqlTypeName.ANY));
          b.add(iRef.accept(new RexVisitor(schema, false, root.getCluster().getRexBuilder())));
```